File: leecode/242. Valid Anagram .js
/*
Given two strings s and t , write a function to determine if t is an anagram of s.

Example 1:

Input: s = "anagram", t = "nagaram"
Output: true
Example 2:

Input: s = "rat", t = "car"
Output: false
Note:
You may assume the string contains only lowercase alphabets.

Follow up:
What if the inputs contain unicode characters? How would you adapt your solution to such case?

*/

/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
// counting sort 的思想, 用 hash 来做
var isAnagram = function(s, t) {
  if (s.length !== t.length) {
    return false
  }
  const count = new Array(26).fill(0)
  // ++ 和 --  一定是等于 0
  for (let i = 0; i < s.length; i++) {
    let a = s[i].charCodeAt() - 97
    let b = t[i].charCodeAt() - 97
    count[a]++
    count[b]--
  }
  for (let n of count) {
    if (n !== 0) {
      return false
    }
  }
  return true
};



var isAnagram1 = function(s, t) {
  s = s.split('').sort().join('')
  t = t.split('').sort().join('')
  return s === t
};

File: python/array/242. Valid Anagram .py
"""
242. Valid Anagram
easy
Given two strings s and t , write a function to determine if t is an anagram of s.

Example 1:

Input: s = "anagram", t = "nagaram"
Output: true
Example 2:

Input: s = "rat", t = "car"
Output: false
Note:
You may assume the string contains only lowercase alphabets.

Follow up:
What if the inputs contain unicode characters? How would you adapt your solution to such case?
"""

"""
思路:
第一个用 obj 来记录次数, 次数++, 检查是否都在 obj 里面, 如果在就次数--, 次数小于 0 的话就说明错了
"""

class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False
        obj = {}
        for char in s:
            if char in obj:
                obj[char] += 1
            else:
                obj[char] = 1
        for char in t:
            if not (char in obj):
                return False
            else:
                obj[char] -= 1
                if obj[char] < 0:
                    return False
        return True
